File: Packages/bmlab/roi_stats.m
% Constants
ROIS = { ...
	'wAngular_L', 'wAngular_R', ...
	'wCingulum_Post_L', 'wCingulum_Post_R', ...
	'wFrontal_Med_Orb_L', 'wFrontal_Med_Orb_R', ...
	'wFrontal_Sup_Medial_L', 'wFrontal_Sup_Medial_R', ...
	'wHippocampus_L', 'wHippocampus_R', ...
	'wParaHippocampal_L', 'wParaHippocampal_R', ...
	'wPrecuneus_L', 'wPrecuneus_R', ...
	'wSupraMarginal_L', 'wSupraMarginal_R' ...
};
TASKS = { 'Rest', 'Rest1', 'Rest2' 'Epso', 'Self', 'Verb' };
TASKS_SHORT = { 'R', 'R1', 'R1' 'M', 'S', 'L' };
SUBJECTS = 1:60;
SUBJECTS([8 15 21 48]) = [];
ALPHA = .05;

% SUBJECT_TASK_ROIS
SUBJECT_TASK_ROIS = {};
for i = 1:60
	sbj = sprintf('TADZ0%02d', i);
	if i <= 46; tasks = [1 4 5 6]; else; tasks = 2:6; end;
	for j = tasks
		fname = sprintf('./Datasets/ROI_Stats/%s_%s.csv', sbj, TASKS{j});
		SUBJECT_TASK_ROIS{i, j} = csvread(fname, 1, 1);
	end
end
SUBJECT_TASK_ROIS(47:60, 1) = SUBJECT_TASK_ROIS(47:60, 3);

% STRENGTH_TABLE
STRENGTH_TABLE = {};
for i = 1:60
	if i <= 46; tasks = [1 4 5 6]; else; tasks = 1:6; end;
	for j = tasks
		r = corr(SUBJECT_TASK_ROIS{i, j}');
		STRENGTH_TABLE{i, j} = (1 + r) ./ (1 - r) / 2;
	end
end

% P_TABLE_TASK and P_TABLE_REST
P_TABLE_TASK = [];
P_TABLE_REST = [];
for roi1 = 1:16
	for roi2 = 1:(roi1-1)
		display(sprintf('== Session %s and %s ==', ROIS{roi1}, ROIS{roi2}));
		table = [];
		for i = 1:60
			if i <= 46; tasks = [1 4 5 6]; else; tasks = 1:6; end;
			for j = tasks
				table(i, j) = STRENGTH_TABLE{i, j}(roi1, roi2);
			end
		end
		table = table(SUBJECTS, :);
		P_TABLE_TASK(roi1, roi2) = anova1(table(:, [1 4 5 6]), [], 'off');
		[~, P_TABLE_REST(roi1, roi2)] = ttest(table(44:end, 2), table(44:end, 3));
	end
end

csvwrite('./Datasets/ROI_Stats/P_TABLE_TASK.csv', P_TABLE_TASK);
csvwrite('./Datasets/ROI_Stats/P_TABLE_REST.csv', P_TABLE_REST);
P_TABLE_TASK = csvread('./Datasets/ROI_Stats/P_TABLE_TASK.csv');
P_TABLE_REST = csvread('./Datasets/ROI_Stats/P_TABLE_REST.csv');

% Display p tables with Bonferroni correction
for p_table = { P_TABLE_TASK, P_TABLE_REST }
	hold on
	imagesc(-log10(p_table(end:-1:2, :)))
	xlim([1 16] - .5)
	ylim([1 16] - .5)
	caxis([0 -log10(ALPHA/2/120)])
	colormap hot
	h = colorbar;
	set(get(h, 'title'), 'string', '-log10(p)');
	ax = gca;
	ax.XTick = 1:15;
	ax.YTick = 1:15;
	ax.XTickLabel = strrep(strrep(ROIS(1:15), '_', ' '), 'w', '');
	ax.YTickLabel = strrep(strrep(ROIS(end:-1:2), '_', ' '), 'w', '');
	ax.XTickLabelRotation = 45;
	for x = find(p_table(end:-1:2, :) < ALPHA/2/120 & p_table(end:-1:2, :) ~= 0)
		plot(floor((x-1)/15)+1, mod(x-1, 15)+1, 'k*')
	end
	hold off
end

% Post-hoc pairwise comparison
for roi1 = 1:16
	for roi2 = 1:(roi1-1)
		if P_TABLE_TASK(roi1, roi2) > ALPHA/2/120
			continue
		end
		roi1_str = strrep(strrep(ROIS{roi1}, '_', ' '), 'w', '');
		roi2_str = strrep(strrep(ROIS{roi2}, '_', ' '), 'w', '');
		display(sprintf('== Session %s and %s ==', ROIS{roi1}, ROIS{roi2}));
		table = [];
		for i = 1:60
			if i <= 46; tasks = [1 4 5 6]; else; tasks = 1:6; end;
			for j = tasks
				table(i, j) = STRENGTH_TABLE{i, j}(roi1, roi2);
			end
		end
		table = table(SUBJECTS, [1 4 6 5]);
		main = sprintf('FC of %s and %s', roi1_str, roi2_str);
		fname = sprintf('./Datasets/ROI_Stats/%s_and_%s.png', ROIS{roi1}, ROIS{roi2});
		stats_task(table, main, fname);
	end
end

File: Projects/bmlab/roi_stats.R
library(oro.nifti)
library(psych)

ROIS <- c(
	"wAngular_L", "wAngular_R",
	"wCingulum_Post_L", "wCingulum_Post_R",
	"wFrontal_Med_Orb_L", "wFrontal_Med_Orb_R",
	"wFrontal_Sup_Medial_L", "wFrontal_Sup_Medial_R",
	"wHippocampus_L", "wHippocampus_R",
	"wParaHippocampal_L", "wParaHippocampal_R",
	"wPrecuneus_L", "wPrecuneus_R",
	"wSupraMarginal_L", "wSupraMarginal_R"
) #=> ROIS

## TASK_SUBJECT_ROIS
TASK_SUBJECT_ROIS <- vector("list", 6)
names(TASK_SUBJECT_ROIS) <- c("Rest", "Rest1", "Rest2", "Epso", "Self", "Verb")
for (task in names(TASK_SUBJECT_ROIS)) {
	TASK_SUBJECT_ROIS[[task]] <- vector("list", 60)
	names(TASK_SUBJECT_ROIS[[task]]) <- sprintf("TADZ0%02d", 1:60)
	for (i in 1:60) {
		TASK_SUBJECT_ROIS[[task]][[sprintf("TADZ0%02d", i)]] <- matrix(0, 16, 211)
		rownames(TASK_SUBJECT_ROIS[[task]][[sprintf("TADZ0%02d", i)]]) <- ROIS
	} #=> for
} #=> for

for (i in 1:60) {
	sbj <- sprintf("TADZ0%02d", i)
	tasks <- c(if (i <= 46) "Rest" else c("Rest1", "Rest2"), "Epso", "Self", "Verb")
	for (task in tasks) {
		fname <- sprintf("./Databank/TADZ0%02d/RESULT/FCA_%s_detrend_filtered/Filtered_4DVolume.nii", i, task)
		img <- matrix(readNIfTI(fname, reorient = FALSE), ncol = 211L)
		for (roi in ROIS) {
			cat(sprintf("== Session %s > %s > %s ==\n", sbj, task, roi))
			roifname <- sprintf("./Datasets/DMN_Masks/TADZ0%02d/m%s.nii", i, roi)
			roiimg <- readNIfTI(roifname, reorient = FALSE)
			roimask <- which(roiimg@.Data == 1)
			TASK_SUBJECT_ROIS[[task]][[sbj]][roi, ] <- colMeans(img[roimask, ])
		} #=> for
		write.csv(TASK_SUBJECT_ROIS[[task]][[sbj]], sprintf("./Datasets/ROI_Stats/%s_%s.csv", sbj, task), quote = FALSE)
	} #=> for
} #=> for

for (i in 1:60) {
	sbj <- sprintf("TADZ0%02d", i)
	tasks <- c(if (i <= 46) "Rest" else c("Rest1", "Rest2"), "Epso", "Self", "Verb")
	for (task in tasks) {
		TASK_SUBJECT_ROIS[[task]][[sbj]] <-
		read.csv(sprintf("./Datasets/ROI_Stats/%s_%s.csv", sbj, task), row.names = 1)
	} #=> for
} #=> for
for (i in 47:60) {
	sbj <- sprintf("TADZ0%02d", i)
	TASK_SUBJECT_ROIS[["Rest"]][[sbj]] <- TASK_SUBJECT_ROIS[["Rest2"]][[sbj]]
} #=> for

## STRENGTH_TABLE
STRENGTH_TABLE <- vector("list", 6)
names(STRENGTH_TABLE) <- c("Rest", "Rest1", "Rest2", "Epso", "Self", "Verb")
for (task in c("Rest", "Epso", "Self", "Verb")) {
	STRENGTH_TABLE[[task]] <- vector("list", 60)
	names(STRENGTH_TABLE[[task]]) <- sprintf("TADZ0%02d", 1:60)
	for (i in 1:60) {
		sbj <- sprintf("TADZ0%02d", i)
		STRENGTH_TABLE[[task]][[sbj]] <- fisherz(cor(t(TASK_SUBJECT_ROIS[[task]][[sbj]])))
	} #=> for
} #=> for
for (task in c("Rest1", "Rest2")) {
	STRENGTH_TABLE[[task]] <- vector("list", 14)
	names(STRENGTH_TABLE[[task]]) <- sprintf("TADZ0%02d", 47:60)
	for (i in 47:60) {
		sbj <- sprintf("TADZ0%02d", i)
		STRENGTH_TABLE[[task]][[sbj]] <- fisherz(cor(t(TASK_SUBJECT_ROIS[[task]][[sbj]])))
	} #=> for
} #=> for

## T_TABLE and SIG_TABLE
T_TABLE <- vector("list", 16)
names(T_TABLE) <- ROIS
SIG_TABLE <- vector("list", 4)
names(SIG_TABLE) <- c("Rest", "Epso", "Self", "Verb")
for (i in 1:length(SIG_TABLE)) {
	SIG_TABLE[[i]] <- matrix(0, 16, 16)
	colnames(SIG_TABLE[[i]]) <- rownames(SIG_TABLE[[i]]) <- ROIS
} #=> for
for (roi1 in 1:16) {
	T_TABLE[[roi1]] <- vector("list", 16)
	names(T_TABLE[[roi1]]) <- ROIS
	for (roi2 in 1:16) {
		if (roi1 == roi2)
			next
		T_TABLE[[roi1]][[roi2]] <- matrix(0, 4, 4)
		rownames(T_TABLE[[roi1]][[roi2]]) <- colnames(T_TABLE[[roi1]][[roi2]]) <- c("Rest", "Epso", "Self", "Verb")
		table <- matrix(0, 60, 4)
		colnames(table) <- c("Rest", "Epso", "Self", "Verb")
		for (task in colnames(table)) {
			for (i in 1:60) {
				table[i, task] <- STRENGTH_TABLE[[task]][[sprintf("TADZ0%02d", i)]][roi1, roi2]
			} #=> for
		} #=> for
		for (i in 1:4) {
			for (j in 1:4) {
				T_TABLE[[roi1]][[roi2]][i, j] <- t.test(table[, i], table[, j], paired = TRUE)$p.value
			} #=> for
		} #=> for
		for (i in 2:4) {
			if (T_TABLE[[roi1]][[roi2]][i] < 0.05 / 120 / 4) {
				SIG_TABLE[[i]][roi1, roi2] <- ifelse(sum(table[, i] - table[, 1]) > 0, 1, -1)
			} #=> if
		} #=> for
	} #=> for
} #=> for
for (roi1 in 1:16) {
	for (roi2 in 1:16) {
		if (roi1 == roi2)
			next
		table <- matrix(0, 14, 2)
		colnames(table) <- c("Rest1", "Rest2")
		for (task in colnames(table)) {
			for (i in 47:60) {
				table[i - 46, task] <- STRENGTH_TABLE[[task]][[sprintf("TADZ0%02d", i)]][roi1, roi2]
			} #=> for
		} #=> for
		SIG_TABLE[[1]][roi1, roi2] <- t.test(table[, 1], table[, 2], paired = TRUE)$p.value < 0.05 / 120 / 4
	} #=> for
} #=> for

## Display tables
library(ggplot2)
require(reshape2)

for (task in names(SIG_TABLE)) {
	plt <- ggplot(data = melt(SIG_TABLE[[task]]), aes(Var1, Var2)) +
	 	geom_tile(aes(fill = factor(value, levels = c("-1", "0", "1"))), color = "black") +
		scale_fill_manual(values = c("red", "white", "green"), drop = FALSE) +
		theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
		xlab("") + ylab("")
	plt$labels$fill <- "Significant"
	plt
} #=> for
